File: 08-安全/02-网络安全/TLS-ECH/go/ech_config.go
// ech_config.go —— ECHConfig 与 ECHClientHello 的编解码（RFC 9849 §4.1、§5）
//
// ECHConfig 是服务端通过 HTTPS DNS 记录（或 DNS 里的 ech 参数）发布会话密钥的部分。
// 注意 **KEM 公钥与套件列表都带长度前缀**，而 maximum_name_length / public_name 是
// u8 前缀 —— 少一层的写法在本地自测里完全正常，只有和真实配置对拍才会暴露。
//
// ECHClientHello（ECH 扩展的内容）只有 outer 一种变体需要序列化：inner 变体是 Empty，
// 所以解码函数对非 0 的首字节一律拒绝。
package main

import "errors"

// ---------------------------------------------------------------- ECHConfig（§4.1）

type hpkeKeyConfig struct {
	ConfigID     byte
	KemID        uint16
	PublicKey    []byte
	CipherSuites [][2]uint16
}

type echConfig struct {
	KeyConfig     hpkeKeyConfig
	MaxNameLength byte
	PublicName    []byte
	Version       uint16
}

func (c *echConfig) Encode() ([]byte, error) {
	w := &wbuf{}
	w.u16(echVersion)
	at := w.vec16Start()
	w.u8(c.KeyConfig.ConfigID)
	w.u16(c.KeyConfig.KemID)
	w.u16(uint16(len(c.KeyConfig.PublicKey)))
	w.put(c.KeyConfig.PublicKey)
	inner := w.vec16Start()
	for _, s := range c.KeyConfig.CipherSuites {
		w.u16(s[0])
		w.u16(s[1])
	}
	if err := w.vec16End(inner); err != nil {
		return nil, err
	}
	w.u8(c.MaxNameLength)
	w.u8(byte(len(c.PublicName)))
	w.put(c.PublicName)
	inner = w.vec16Start() // extensions：本 demo 恒空
	if err := w.vec16End(inner); err != nil {
		return nil, err
	}
	if err := w.vec16End(at); err != nil {
		return nil, err
	}
	return w.b, nil
}

// decodeEchConfig —— 只接受本 demo 支持的版本与「无扩展」的 contents；
// 真实实现里不认识的版本要靠 length 跳过（§4.2），见 README 的跨语言一致性一节
func decodeEchConfig(data []byte) (*echConfig, error) {
	r := newRbuf(data)
	ver, err := r.u16()
	if err != nil {
		return nil, err
	}
	if ver != echVersion {
		return nil, errors.New("ech: 不认识的 ECHConfig 版本")
	}
	body, err := r.vec16()
	if err != nil || !r.done() {
		return nil, errShort
	}
	b := newRbuf(body)
	c := &echConfig{Version: ver}
	if c.KeyConfig.ConfigID, err = b.u8(); err != nil {
		return nil, err
	}
	if c.KeyConfig.KemID, err = b.u16(); err != nil {
		return nil, err
	}
	if c.KeyConfig.PublicKey, err = b.vec16(); err != nil || len(c.KeyConfig.PublicKey) == 0 {
		return nil, errShort
	}
	suites, err := b.vec16()
	if err != nil || len(suites) == 0 || len(suites)%4 != 0 {
		return nil, errShort
	}
	for i := 0; i < len(suites); i += 4 {
		c.KeyConfig.CipherSuites = append(c.KeyConfig.CipherSuites,
			[2]uint16{uint16(suites[i])<<8 | uint16(suites[i+1]),
				uint16(suites[i+2])<<8 | uint16(suites[i+3])})
	}
	if c.MaxNameLength, err = b.u8(); err != nil {
		return nil, err
	}
	if c.PublicName, err = b.vec8(); err != nil || len(c.PublicName) == 0 {
		return nil, errShort
	}
	exts, err := b.vec16()
	if err != nil || !b.done() {
		return nil, errShort
	}
	if len(exts) != 0 {
		return nil, errors.New("ech: 本实现不支持带扩展的 ECHConfig")
	}
	return c, nil
}

// ---------------------------------------------------------------- ECHClientHello（§5）

type outerExt struct {
	KdfID    uint16
	AeadID   uint16
	ConfigID byte
	Enc      []byte
	Payload  []byte
}

func encodeOuterExt(o *outerExt) ([]byte, error) {
	w := &wbuf{}
	w.u8(0x00) // ECHClientHelloType = outer(0)
	w.u16(o.KdfID)
	w.u16(o.AeadID)
	w.u8(o.ConfigID)
	w.u16(uint16(len(o.Enc)))
	w.put(o.Enc)
	w.u16(uint16(len(o.Payload)))
	w.put(o.Payload)
	return w.b, nil
}

func decodeOuterExt(data []byte) (*outerExt, error) {
	r := newRbuf(data)
	typ, err := r.u8()
	if err != nil {
		return nil, err
	}
	if typ != 0x00 {
		return nil, errors.New("ech: 不是 outer 变体")
	}
	o := &outerExt{}
	if o.KdfID, err = r.u16(); err != nil {
		return nil, err
	}
	if o.AeadID, err = r.u16(); err != nil {
		return nil, err
	}
	if o.ConfigID, err = r.u8(); err != nil {
		return nil, err
	}
	if o.Enc, err = r.vec16(); err != nil {
		return nil, err
	}
	if o.Payload, err = r.vec16(); err != nil || !r.done() {
		return nil, errShort
	}
	return o, nil
}

File: 08-安全/02-网络安全/TLS-ECH/python/ech_config.py
"""
ECHConfig / ECHConfigList / ECHClientHello 的编解码（RFC 9849 §4-§5）

ECHConfig 从 DNS 的 HTTPS 记录里来，是**不可信输入**；这里把每条长度前缀都做了检查，
并把规范里「必须忽略/必须 abort」的几种情况实现成显式异常：
  * 不认识的 version  -> 靠 length 字段跳过（§4）
  * 重复的扩展         -> 拒绝整条 ECHConfig（§4.2）
  * 高位为 1 的强制扩展 -> 拒绝整条 ECHConfig（§4.2）
"""

from __future__ import annotations

# 从 ech_wire 转出（本仓库内部模块，调用方可以只 import ech_config）
from ech_wire import (AEAD_ID_CHACHA20POLY1305, ECH_EXT_TYPE, ECH_OUTER_EXTENSIONS,  # noqa: F401
                      ECH_VERSION, EXT_KEY_SHARE, EXT_SERVER_NAME,               # noqa: F401
                      EXT_SUPPORTED_VERSIONS, KDF_ID_HKDF_SHA256, KEM_ID_X25519,  # noqa: F401
                      ClientHello, EchError, Reader, u8, u16, vec8, vec16)        # noqa: F401

# ------------------------------------------------------------ ECHConfig（RFC 9849 §4）

class HpkeKeyConfig:
    def __init__(self, config_id: int, kem_id: int, public_key: bytes,
                 cipher_suites) -> None:
        self.config_id = config_id
        self.kem_id = kem_id
        self.public_key = public_key
        self.cipher_suites = list(cipher_suites)      # [(kdf_id, aead_id), ...]

    def encode(self) -> bytes:
        suites = b"".join(u16(k) + u16(a) for k, a in self.cipher_suites)
        return u8(self.config_id) + u16(self.kem_id) + vec16(self.public_key) + vec16(suites)

    @classmethod
    def decode(cls, r: Reader) -> "HpkeKeyConfig":
        config_id = r.u8()
        kem_id = r.u16()
        public_key = r.opaque16()
        blob = r.opaque16()
        if len(blob) < 4 or len(blob) % 4:
            raise EchError("cipher_suites 长度必须是 4 的倍数且至少 4 字节")
        suites = [(int.from_bytes(blob[i:i + 2], "big"),
                   int.from_bytes(blob[i + 2:i + 4], "big"))
                  for i in range(0, len(blob), 4)]
        return cls(config_id, kem_id, public_key, suites)


class EchConfig:
    def __init__(self, key_config: HpkeKeyConfig, maximum_name_length: int,
                 public_name: bytes, extensions=(), version: int = ECH_VERSION) -> None:
        self.version = version
        self.key_config = key_config
        self.maximum_name_length = maximum_name_length
        self.public_name = public_name
        self.extensions = list(extensions)             # [(type, data), ...]

    def contents(self) -> bytes:
        exts = b"".join(u16(t) + vec16(d) for t, d in self.extensions)
        return (self.key_config.encode() + u8(self.maximum_name_length)
                + vec8(self.public_name) + vec16(exts))

    def encode(self) -> bytes:
        return u16(self.version) + vec16(self.contents())

    @classmethod
    def decode(cls, data: bytes) -> "EchConfig":
        r = Reader(data)
        cfg = cls._decode_from(r)
        r.expect_end()
        return cfg

    @classmethod
    def _decode_from(cls, r: Reader) -> "EchConfig":
        """从「version 字段」处开始读一条 ECHConfig（ECHConfigList 里连续排列）。"""
        version = r.u16()
        if version != ECH_VERSION:
            raise EchError(f"不支持的 ECH 版本 0x{version:04x}")
        body = Reader(r.opaque16())
        key_config = HpkeKeyConfig.decode(body)
        maximum_name_length = body.u8()
        public_name = body.opaque8()
        if not public_name:
            raise EchError("public_name 不能为空")
        ext_blob = body.opaque16()
        body.expect_end()
        er = Reader(ext_blob)
        extensions = []
        seen = set()
        while er.pos < len(ext_blob):
            etype = er.u16()
            edata = er.opaque16()
            if etype in seen:
                raise EchError("ECHConfig 扩展重复（RFC 9849 §4.2 禁止）")
            if etype & 0x8000:
                raise EchError(f"不认识的强制扩展 0x{etype:04x}，必须忽略整条 ECHConfig")
            seen.add(etype)
            extensions.append((etype, edata))
        return cls(key_config, maximum_name_length, public_name, extensions, version)


def encode_config_list(configs) -> bytes:
    return vec16(b"".join(c.encode() for c in configs))


def decode_config_list(data: bytes) -> list:
    """RFC 9849 §4：不认识的 version **靠 length 字段跳过**，而不是让整条列表失败。

    这正是 ECHConfig 里那个看似冗余的 length 字段存在的唯一理由 —— 客户端升级期
    必须能忽略未来的新版本，继续用列表里自己支持的那些配置。
    """
    r = Reader(data)
    blob = r.opaque16()
    r.expect_end()
    ir = Reader(blob)
    out = []
    while ir.pos < len(blob):
        version = int.from_bytes(ir.data[ir.pos:ir.pos + 2], "big")
        if version != ECH_VERSION:
            ir.u16()
            ir.opaque16()                      # 跳过这条
            continue
        out.append(EchConfig._decode_from(ir))
    return out


# ------------------------------------------------------------ ECHClientHello（RFC 9849 §5）

def ech_inner_extension() -> bytes:
    """inner 变体：只有一个 type 字节。存在的意义只是「服务端不许回没提议过的扩展」。"""
    return b"\x01"


def ech_outer_extension(cipher_suite, config_id: int, enc: bytes, payload: bytes) -> bytes:
    kdf_id, aead_id = cipher_suite
    return (b"\x00" + u16(kdf_id) + u16(aead_id) + u8(config_id)
            + vec16(enc) + vec16(payload))


def parse_ech_outer_extension(data: bytes):
    """返回 (cipher_suite, config_id, enc, payload)。"""
    r = Reader(data)
    if r.u8() != 0x00:
        raise EchError("不是 outer 变体的 ECH 扩展")
    suite = (r.u16(), r.u16())
    config_id = r.u8()
    enc = r.opaque16()
    payload = r.opaque16()
    r.expect_end()
    return suite, config_id, enc, payload